netket/optimizer/jax/_sr_onthefly.py

```diff
@@ -0,0 +1,92 @@
+import jax
+import jax.numpy as jnp
+from functools import partial
+
+
+# onthefly SR
+# since we cant store O if #samples x #params is too large
+
+
+def O_jvp(x, theta, v, vlogwf):
+    _, res = jax.jvp(lambda p: vlogwf(p, x), (theta,), (v,))
+    return res
+
+
+def O_vjp(x, theta, v, vlogwf, return_vjp_fun=False, vjp_fun=None):
+    if vjp_fun is None:
+        _, vjp_fun = jax.vjp(vlogwf, theta, x)
+    res, _ = vjp_fun(v)
+    if return_vjp_fun:
+        return res, vjp_fun
+    else:
+        return res
+
+
+# calculate \langle O \rangle
+def Obar(samples, theta, vlogwf, **kwargs):
+    # TODO better way to get dtype
+    dtype = vlogwf(theta, samples[:1])[0].dtype
+    v = jnp.ones(samples.shape[0], dtype=dtype) / samples.shape[0]
+    return O_vjp(samples, theta, v, vlogwf, **kwargs)
+
+
+# calculate O^\dagger O v
+# @partial(jax.jit, static_argnums=3)
+def odagov(samples, theta, v, vlogwf):
+    vprime = O_jvp(samples, theta, v, vlogwf)
+    res = O_vjp(samples, theta, vprime.conjugate(), vlogwf)
+    return jax.tree_map(jax.lax.conj, res)  # return res.conjugate()
+
+
+# calculate O^\dagger \Delta O v
+# where \Delta O = O-\langle O \rangle
+# optional: pass jvp_fun to be reused
+# TODO vjp_fun and jit??
+# @partial(jax.jit, static_argnums=3)
+def odagdeltaov(samples, theta, v, vlogwf, vjp_fun=None, factor=1.0):
+    # reuse vjp_fun from O_mean below for O_vjp
+    O_mean, vjp_fun = Obar(samples, theta, vlogwf, return_vjp_fun=True, vjp_fun=vjp_fun)
+    vprime = O_jvp(samples, theta, v, vlogwf)  # is an array of size n_samples
+    # TODO tree_dot would be nice
+    # here we use jax.numpy.add to automatically promote nonhomogeneous parameters to the larger type
+
+    # omeanv = O_mean.dot(v); is a scalar
+    omeanv = jax.tree_util.tree_reduce(
+        jax.numpy.add,
+        jax.tree_map(jax.numpy.sum, jax.tree_multimap(jax.lax.mul, O_mean, v)),
+    )
+    # vprime -= omeanv (elementwise)
+    vprime = vprime - jax.lax.broadcast(omeanv, vprime.shape)
+    # vprime *= factor (elementwise)
+    vprime = jax.lax.mul(
+        vprime, jax.lax.broadcast(jnp.array(factor, dtype=vprime.dtype), vprime.shape)
+    )
+
+    res = O_vjp(samples, theta, vprime.conjugate(), vlogwf, vjp_fun=vjp_fun)
+    res = jax.tree_map(jax.lax.conj, res)  # res = res.conjugate()
+    # convert back the parameters which were promoted earlier:
+    # astype alone would also work, however this raises ComplexWarning when casting complex to real, so we take the real where needed first
+    res = jax.tree_multimap(
+        lambda x, target: (x if jnp.iscomplexobj(target) else x.real).astype(
+            target.dtype
+        ),
+        res,
+        v,
+    )
+    return res
+
+
+def mat_vec(v, forward_fn, params, samples, diag_shift, n_samp):
```

Comment:
n_samp is a redundant parameter? And it's static when jitting so even better.

```suggestion
def mat_vec(v, forward_fn, params, samples, diag_shift):
    n_samp = samples.shape[0]
```

Reply:
I actually kept this so that if we want to parallelize it it with MPI at some point,
then we can divide by the n_samp across all MPI processes and just sum (via allreduce) at the end.
(in analogy to the matrix version)

What do you think?

Reply:
Actually I think all that is missing for MPI is to add the linear transpose rule (ad.primitive_transposes) for mpi_allreduce_p in mpi4jax, (or use jax._src.scipy.sparse.linalg._cg_solve directly, which does not require it; hardcoding symmetric=True in jax._src.scipy.sparse.linalg.cg is also a possible workaround just for testing it)

Reply:
The number of MPI nodes is a known and runtime constant, so you don't need it?

As for the rule... yes, IIRC I [have defined](https://github.com/PhilipVinc/mpi4jax/blob/875e60b1737014ecd6885f4c64564705b5ce3c81/mpi4jax/collective_ops/allreduce.py#L157) the backward jvp pass but not the forward vip pass on allreduce. 
 It should be trivial to add